### ISSUE TITLE
nosetests and pytest hang after test completion because of leaking resources in SNS

### ISSUE THREAD

**Issue by alanyang3** [localstack/localstack#209]
Minimum code running on OSX, python 3.6:
```
import os
import json
import uuid
from localstack.services import infra
from localstack.utils.aws import aws_stack

class TestPredictionStream(object):
    LOCALSTACK_HOST = 'http://localhost'
    SNS_PORT = '4575'
    SQS_PORT = '4576'
    SQS_ENDPOINT_URL = LOCALSTACK_HOST + ':' + SQS_PORT
    SNS_ENDPOINT_URL = LOCALSTACK_HOST + ':' + SNS_PORT

    TEST_QUEUE_NAME = 'test_queue'
    TEST_SNS_TOPIC = 'test_topic'

    @classmethod
    def setup_class(cls):
        services_needed_for_test = ','.join([
            'sqs:' + TestPredictionStream.SQS_PORT,
            'sns:' + TestPredictionStream.SNS_PORT
        ])
        os.environ['SERVICES'] = services_needed_for_test
        infra.start_infra(async=True)

    @classmethod
    def teardown_class(cls):
        infra.stop_infra()

    def setup(self):
        self.sqs_client = aws_stack.connect_to_service('sqs')
        self.sns_client = aws_stack.connect_to_service('sns')
        self.queue_info = self._setup_queue()
        self.topic_info = self._setup_topic()

    def _setup_queue(self):
        self.sqs_client.create_queue(QueueName=self.TEST_QUEUE_NAME)

    def _setup_topic(self):
        return self.sns_client.create_topic(Name=self.TEST_SNS_TOPIC)

    def teardown(self):
        self._teardown_queue()
        self._teardown_topic()

    def _teardown_queue(self):
        self.sqs_client.delete_queue(QueueUrl=self.TEST_QUEUE_NAME)

    def _teardown_topic(self):
        self.sns_client.delete_topic(TopicArn=self.topic_info['TopicArn'])

    def test_something(self):
        # stub test
        pass
```

Tests finish running, but the nose and pytest runner will never exit. Seems to be okay with unittest.

```
Starting mock SNS (http port 4575)...
Starting mock SQS (http port 4576)...
Ready.
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): 127.0.0.1
requests.packages.urllib3.connectionpool: DEBUG: "POST / HTTP/1.1" 200 337
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): 127.0.0.1
requests.packages.urllib3.connectionpool: DEBUG: "POST / HTTP/1.1" 200 200
.
----------------------------------------------------------------------
Ran 1 test in 4.467s

OK
<never exits>
```

Debugging in to the test runner showed that it was stuck on a thread join:
```
> /usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py(1294)_shutdown()
-> t.join()
```

Problem goes away if SNS topic is never created

**Comment by alanyang3** [comment]
Should also add that netstat shows the SNS port is still established after the infra.stop_infra() is called:

```
ps

  PID TTY           TIME CMD
  478 ttys000    0:00.63 -zsh
 8044 ttys000    0:04.44 /Users/alanyang/.virtualenvs/env/bin/python /Users/alanyang/.virtualenvs/env/bin/nosetests -s
 1688 ttys001    0:00.38 -zsh

netstat -an | grep 4575
tcp4       0      0  127.0.0.1.4575         127.0.0.1.50602        ESTABLISHED
tcp4       0      0  127.0.0.1.50602        127.0.0.1.4575         ESTABLISHED
```

**Comment by whummer** [comment]
Thanks for the detailed report @alanyang3 . I was able to reproduce the issue locally. This should be  fixed in https://github.com/localstack/localstack/pull/211. Can you please give it a try with the latest version of the code? Thanks

**Comment by alanyang3** [comment]
No problem. Thank you for the quick response!

#211 does fix it. Will you be able to release to pip soon?

**Comment by whummer** [comment]
Done, released as `0.7.2`.

**Comment by alanyang3** [comment]
Awesome, well done
